File: Main-c.cs
// Bradford Arrington 2025
using System;
using System.IO;
using System.Collections.Generic;

class Program
{
    public static void Main(string[] args)
    {
		using (new AdvancedPerformanceMonitor("Detail Operations"))
		{
			var compressor = new Compressor();
			string[] arguments = Environment.GetCommandLineArgs();
			if (args.Length < 2)
			{
			   string progName = arguments[0];
			   string shortName = progName;
			   int lastSlash = progName.LastIndexOf('\\');
			   if (lastSlash == -1)
					lastSlash = progName.LastIndexOf('/');
			   if (lastSlash == -1)
					lastSlash = progName.LastIndexOf(':');
			   if (lastSlash != -1)
					shortName = progName.Substring(lastSlash + 1);
			   int extension = shortName.LastIndexOf('.');
			   if (extension != -1)
					shortName = shortName.Substring(0, extension);
			   Console.WriteLine($"\nUsage:  {shortName} {Compressor.Usage}");
			   Environment.Exit(0);
			}
			string[] remainingArgs = new string[args.Length - 2];
			Array.Copy(args, 2, remainingArgs, 0, remainingArgs.Length);
			Compressor.BitFile output;
			FileStream input;
			using (new TimeMonitor("ReadFile"))
			{
				Compressor.BitFile output = Compressor.BitFile.OpenOutputBitFile(args[1]);
				FileStream input = new FileStream(args[0], FileMode.Open, FileAccess.Read);
			}
			Console.WriteLine($"\nCompressing {args[0]} to {args[1]}");
			Console.WriteLine($"Using {compressor.CompressionName}\n");
		    using (new TimeMonitor("CompressFile and Close"))
			{
				compressor.CompressFile(input, output, remainingArgs.Length, remainingArgs);
				output.CloseBitFile();
				input.Close();
			}
			PrintRatios(args[0], args[1]);
		}
    }

    public static long FileSize(string fileName)
    {
        try
        {
            var fileInfo = new FileInfo(fileName);
            return fileInfo.Length;
        }
        catch
        {
            return 0L;
        }
    }

    public static void PrintRatios(string inputFilePath, string outputFilePath)
    {
        long inputSize = FileSize(inputFilePath);
        if (inputSize == 0)
        {
            inputSize = 1;
        }

        long outputSize = FileSize(outputFilePath);
        int ratio = 100 - (int)(outputSize * 100L / inputSize);

        Console.WriteLine($"\nInput bytes:        {inputSize}");
        Console.WriteLine($"Output bytes:       {outputSize}");
        Console.WriteLine($"Compression ratio:  {ratio}%");
    }
}

File: Java/mainc.java
import java.io.*;
import java.util.*;

public class mainc {
    
    public static void main(String[] args) {
        if (args.length < 2) {
            usageExit("lzw12");
        }
        
        try {
            File inputFile = new File(args[0]);
            File outputFile = new File(args[1]);
            
            String[] additionalArgs = new String[0];
            if (args.length > 2) {
                additionalArgs = Arrays.copyOfRange(args, 2, args.length);
            }
            
            System.out.printf("\nCompressing %s to %s\n", args[0], args[1]);
            System.out.println("Using LZW 12 Bit Encoder");
            
            long startTime = System.currentTimeMillis();
            
            try (FileInputStream input = new FileInputStream(inputFile);
                 FileOutputStream output = new FileOutputStream(outputFile)) {
                
                BitOutputStream bitOutput = new BitOutputStream(output);
                lzw12 lzw = new lzw12();
                lzw.compressFile(input, bitOutput, additionalArgs);
                bitOutput.close();
            }
            
            long endTime = System.currentTimeMillis();
            
            printRatios(inputFile, outputFile);
            System.out.printf("Compression time: %d ms\n", endTime - startTime);
            
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    private static void usageExit(String progName) {
        String shortName = progName;
        int lastSlash = Math.max(progName.lastIndexOf('\\'), progName.lastIndexOf('/'));
        if (lastSlash != -1) {
            shortName = progName.substring(lastSlash + 1);
        }
        
        // Remove extension if present
        int dotIndex = shortName.lastIndexOf('.');
        if (dotIndex != -1) {
            shortName = shortName.substring(0, dotIndex);
        }
        
        System.out.printf("\nUsage: %s in-file out-file \n\n", shortName);
        System.exit(0);
    }
    
    private static void printRatios(File inputFile, File outputFile) throws IOException {
        long inputSize = fileSize(inputFile);
        if (inputSize == 0) {
            inputSize = 1;
        }
        
        long outputSize = fileSize(outputFile);
        int ratio = 100 - (int)(outputSize * 100L / inputSize);
        
        System.out.printf("\nInput bytes:        %d\n", inputSize);
        System.out.printf("Output bytes:       %d\n", outputSize);
        if (outputSize == 0) outputSize = 1;
        System.out.printf("Compression ratio:  %d%%\n", ratio);
    }
    
    private static long fileSize(File file) throws IOException {
        return file.length();
    }
}
